src/vizier/services/query_broker/script_runner/config_map_source.go

```diff
@@ -20,111 +20,161 @@ package scriptrunner
+			informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+				options.LabelSelector = "purpose=cron-script"
+			}),
+		).Core().V1().ConfigMaps(),
+	}
 }
 
 // Start watches for updates to matching configmaps and sends resulting updates on updatesCh.
-func (source *ConfigMapSource) Start(baseCtx context.Context, updatesCh chan<- *cvmsgspb.CronScriptUpdate) (map[string]*cvmsgspb.CronScript, error) {
-	options := metav1.ListOptions{LabelSelector: "purpose=cron-script"}
-	watcher, err := source.client.Watch(baseCtx, options)
+func (source *ConfigMapSource) Start(_ context.Context, updatesCh chan<- *cvmsgspb.CronScriptUpdate) (map[string]*cvmsgspb.CronScript, error) {
+	stopCh := make(chan struct{})
+	isInitialized := &atomic.Bool{}
+	_, err := source.informer.Informer().AddEventHandler(configMapEventHandlers(isInitialized, updatesCh))
+	if err != nil {
+		return nil, err
+	}
+	go source.informer.Informer().Run(stopCh)
+	waitUntilSynced(source.informer.Informer())
+	initialScripts, err := getInitialScripts(source.informer.Lister())
 	if err != nil {
+		close(stopCh)
 		return nil, err
 	}
-	go configMapUpdater(watcher, updatesCh)
-	configmaps, err := source.client.List(baseCtx, options)
+	isInitialized.Store(true)
+	source.stop = func() { close(stopCh) }
+	return initialScripts, nil
+}
+
+func getInitialScripts(lister listercorev1.ConfigMapLister) (map[string]*cvmsgspb.CronScript, error) {
+	configMaps, err := lister.List(labels.Everything())
 	if err != nil {
-		watcher.Stop()
 		return nil, err
 	}
 	scripts := map[string]*cvmsgspb.CronScript{}
-	for _, configmap := range configmaps.Items {
-		id, cronScript, err := configmapToCronScript(&configmap)
+	for _, configMap := range configMaps {
+		id, cronScript, err := configmapToCronScript(configMap)
 		if err != nil {
 			logCronScriptParseError(err)
 			continue
 		}
 		scripts[id] = cronScript
 	}
-	source.stop = watcher.Stop
 	return scripts, nil
 }
 
-// Stop stops further updates from being sent.
-func (source *ConfigMapSource) Stop() {
-	source.stop()
-}
-
-func configMapUpdater(watcher watch.Interface, updatesCh chan<- *cvmsgspb.CronScriptUpdate) {
-	for event := range watcher.ResultChan() {
-		switch event.Type {
-		case watch.Modified, watch.Added:
-			configmap := event.Object.(*v1.ConfigMap)
-			id, script, err := configmapToCronScript(configmap)
-			if err != nil {
-				logCronScriptParseError(err)
-				continue
-			}
-			cronScriptUpdate := &cvmsgspb.CronScriptUpdate{
-				Msg: &cvmsgspb.CronScriptUpdate_UpsertReq{
-					UpsertReq: &cvmsgspb.RegisterOrUpdateCronScriptRequest{
-						Script: script,
-					},
-				},
-				RequestID: id,
-				Timestamp: time.Now().Unix(),
+func configMapEventHandlers(isInitialized *atomic.Bool, updatesCh chan<- *cvmsgspb.CronScriptUpdate) cache.ResourceEventHandlerFuncs {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			configmap := obj.(*corev1.ConfigMap)
+			if !isInitialized.Load() {
```

Comment:
Do we need to check `isInitialized` for delete and update funcs too?

Reply:
It will only send add events to build the initial list.

Reply:
So what lead me to the idea is that in k8s 1.27, the [`ResourceEventHandler.OnAdd`](https://pkg.go.dev/k8s.io/client-go@v0.27.2/tools/cache#ResourceEventHandler) method has an extra argument `isInInitialList` while the other methods do not.